include: sys: util: Move macros from util to util_macro

nandojve opened the pull request:
To elaborate more advanced macros for dts related bindings the <sys/util.h> can not be include.  It create garbage at
<board>.dts.pre.tmp file and have path include issues.  This moves all macros from util.h to util_macro.h  This allows that
<sys/util_macro.h> can be include on files like <dts/arm/<manufacturer>/pinctrl_<manufacturer>_<soc>.h.

The issue was raised when try create a macro for pincrtl with a variable length flag list with .

Original motivation: #28645

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
include: sys: util: Move macros from util to util_macro

The <sys/util.h>, in current form, can not be used with DTS as it
contains non-C pre-processor definitions which breake DTS interpreter.
This commit fixes the problem by moving most of preprocessor macros
from util.h to util_macro.h.  Since util_mcaro.h contains only
preprocessor macros, without include dependencies, it can be safely
included in DTS. It is similar way the
<dts/arm/<manufacturer>/pinctrl_<manufacturer>_<soc>.h files are
included.

This fix and align the extern "C" closing brack inside non assembly
block.

The issue was raised when try create a macro for pincrtl with a
variable length flag list.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
rebase and update commit message.

de-nordic commented on include/sys/util_macro.h:
```diff
@@ -0,0 +1,621 @@
+#ifndef ZEPHYR_INCLUDE_SYS_UTIL_MACROS_H_
+#define ZEPHYR_INCLUDE_SYS_UTIL_MACROS_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Shouldn't this be taken out to `util.h` to surround the inclusion of `util_macro.h`?
I do not think that the DTS intermediate file will have `extern "C" {` inserted, because it will not have the `__cplusplus` defined, but if this is supposed to have only DTS interpreter acceptable syntax then this block should be thrown outside.
What do you think @galak?

galak replied:
I think its ok to leave here since the only time __cplusplus should get defined is with a C++ compiler.  So with DTS this will just go away.

de-nordic reviewed (approved):
Looks OK.

de-nordic commented:
One more note. While I think that making `util_macro.h` separate file is general good move, I also think that there could be new file created within `include/dt-bindings` that would be wrapping the `include/sys/util_macro.h`, and that file should be included into dts files rather than `include/sys/util_macro.h`.
I leave this decision to DTS experts, @galak, @mbolivar-nordic what do you think?

carlescufi reviewed (commented):
No objections to this, but I'd like someone else to approve

mbolivar-nordic commented:
> I also think that there could be new file created within include/dt-bindings that would be wrapping the include/sys/util_macro.h, and that file should be included into dts files rather than include/sys/util_macro.h.

I agree with you @de-nordic . Our in-tree DTS files so far only `#include <dt-bindings/foo.h>`. There's no including of other files so far, and this feels a bit cleaner.

nandojve commented:
> > I also think that there could be new file created within include/dt-bindings that would be wrapping the include/sys/util_macro.h, and that file should be included into dts files rather than include/sys/util_macro.h.
> 
> I agree with you @de-nordic . Our in-tree DTS files so far only `#include <dt-bindings/foo.h>`. There's no including of other files so far, and this feels a bit cleaner.

Is there any side effects having include/dt-bindings included by util.h? What happen if we need some specifics that can not be shared with util.h?

mbolivar-nordic commented:
> Is there any side effects having include/dt-bindings included by util.h? What happen if we need some specifics that can not be shared with util.h?

I don't understand the question, sorry.

nandojve commented:
> > Is there any side effects having include/dt-bindings included by util.h? What happen if we need some specifics that can not be shared with util.h?
> 
> I don't understand the question, sorry.

After re-reading I think I made a confusion, sorry. Let's recap:

1) every macro will be moved from util.h to util_macro.h and we will keep **extern "C" {**
2) a new file include/dt-bindings/dt-binding.h should be created
3) dt-binding.h will include /sys/util_macros.h and should be used instead of direct access to util_macro.h

If these are the consensus I'm ready to make the adjustments.

nandojve committed:
include: dt-bindings: Add macros entry point for DTS preprocessor

Introduce <dt-bindings/dt-util.h> file.  It wraps <sys/util_macro.h>
file exposing all macro base definitions to DTS preprocessor.  Once it
contains only prepocessor macros, it can be safely included in DTS,
similar way the <dts/arm/<manufacturer>/pinctrl_<manufacturer>_<soc>.h
files are included.

This provides necessary background to elaborate complex constructions
like variable length macros with zero or more elements.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
Update with last comments.

I would like remind that I align extern "C" block inside non ASM block and some definitions are outside extern "C".
It will be good if someone could give OK as double check that before merge.

If needs more adjustments, no problem, just let me know!

de-nordic commented:
@nandojve you have asked me to look again. I think it is OK and uphold my approval.

galak commented on include/dt-bindings/dt-binding.h:
```diff
@@ -0,0 +1,12 @@
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_DT_BINDING_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_DT_BINDING_H_
+
```
Add a comment about why we have this include file just including <sys/util_macro.h>

galak reviewed (changes requested):
rename `dt-binding.h` to something like `dt-util.h`.
